File: typescript/test/smoke.test.ts
/**
 * Env-gated live smoke test (DESIGN.md §9): set WRIT_SMOKE=1 to run against a
 * real, discovered daemon. Read-only — status + two list calls. Skipped by
 * default.
 */

import { describe, expect, it } from "vitest";
import { WritAgent } from "../src/index.js";

const smoke = process.env.WRIT_SMOKE === "1";

describe.skipIf(!smoke)("live smoke (WRIT_SMOKE=1)", () => {
  it("discovers the daemon and answers status + workflows + runs", async () => {
    const client = await WritAgent.discover();

    const status = await client.agent.status();
    expect(status.status).toBe("ok");
    expect(typeof status.version).toBe("string");

    const workflows = await client.workflows.list();
    expect(Array.isArray(workflows.data)).toBe(true);
    expect(workflows.count).toBe(workflows.data.length);

    const runs = await client.runs.list({ limit: 1 });
    expect(Array.isArray(runs.data)).toBe(true);
    expect(runs.data.length).toBeLessThanOrEqual(1);
    expect(runs.total === null || typeof runs.total === "number").toBe(true);
  });
});

File: go/smoke_test.go
package writ

import (
	"context"
	"net/url"
	"os"
	"strconv"
	"testing"
	"time"
)

// TestLiveSmoke is an env-gated smoke test against a real local daemon:
//
//	WRIT_SMOKE=1 go test -run TestLiveSmoke ./...
//
// It discovers the daemon like any client would and performs READ-ONLY
// calls: agent status, workflow list, and a one-item run list.
func TestLiveSmoke(t *testing.T) {
	if os.Getenv("WRIT_SMOKE") != "1" {
		t.Skip("set WRIT_SMOKE=1 to run the live smoke test against a local daemon")
	}
	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()

	c, err := Discover(ctx)
	if err != nil {
		t.Fatalf("Discover: %v", err)
	}

	status, err := c.Agent.Status(ctx)
	if err != nil {
		t.Fatalf("Agent.Status: %v", err)
	}
	if status.Status == "" || status.Version == "" {
		t.Errorf("suspicious agent status: %+v", status)
	}
	t.Logf("agent %s (v%s), active_runs=%d", status.Status, status.Version, status.ActiveRuns)

	workflows, err := c.Workflows.List(ctx, nil)
	if err != nil {
		t.Fatalf("Workflows.List: %v", err)
	}
	if workflows.Count != len(workflows.Data) {
		t.Errorf("count %d != len(data) %d", workflows.Count, len(workflows.Data))
	}
	t.Logf("%d workflows", workflows.Count)

	runs, err := c.Runs.List(ctx, url.Values{"limit": {"1"}})
	if err != nil {
		t.Fatalf("Runs.List: %v", err)
	}
	if len(runs.Data) > 1 {
		t.Errorf("limit=1 returned %d rows", len(runs.Data))
	}
	if len(runs.Data) == 1 {
		if _, ok := runs.Data[0].RowID(); !ok {
			t.Errorf("run feed id %q has no numeric row id", runs.Data[0].ID)
		}
	}
	// Total is *int (nil when the daemon's envelope omits it — see Page). %v on a
	// pointer prints its ADDRESS, so this line used to read `total=0xc000123456`,
	// which looks like a plausible-but-enormous count. The smoke log is the
	// evidence a maintainer reads to decide the run was healthy; it has to say
	// what the daemon actually reported.
	total := "absent"
	if runs.Total != nil {
		total = strconv.Itoa(*runs.Total)
	}
	t.Logf("runs page: count=%d total=%s", runs.Count, total)
}
